File: javascript/leetcode/268. Missing Number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var missingNumber = function(nums) {
  let pos = 0;
  let max = nums[pos];
  const set = new Set();

  while (pos < nums.length) {
    if (nums[pos] > max) max = nums[pos];
    set.add(nums[pos]);
    pos++;
  }

  for (let i = 0; i <= max; i++) {
    if (!set.has(i)) return i;
  }

  return max + 1;
};

File: go/leetcode/268. Missing Number.go
func missingNumber(nums []int) int {
    max := nums[0]
    mp := make(map[int]int)
    
    for i := 0; i < len(nums); i++ {
        mp[nums[i]] = i
        if nums[i] > max {
            max = nums[i]
        }
    }
    
    for i := 0; i <= max; i++ {
        _, ok := mp[i]
        if !ok {
            return i
        }
    }
    
    return max + 1
}